pkg/testing/tools/artifacts_api.go

```diff
@@ -0,0 +1,246 @@
+		return
+	}
+	list = new(VersionList)
+	err = json.Unmarshal(respBytes, list)
+
+	if err != nil {
+		err = fmt.Errorf("unmarshaling version list: %w", err)
+		return
+	}
+
+	return list, nil
+}
+
+func (aac ArtifactAPIClient) GetBuildsForVersion(ctx context.Context, version string) (builds *VersionBuilds, err error) {
+	joinedURL, err := url.JoinPath(aac.apiURL, fmt.Sprintf(artifactsAPIV1VersionBuildsEndpoint, version))
+	if err != nil {
+		err = fmt.Errorf("composing URL with %q %q: %w", aac.apiURL, fmt.Sprintf(artifactsAPIV1VersionBuildsEndpoint, version), err)
+		return
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, joinedURL, nil)
+
+	if err != nil {
+		err = fmt.Errorf("composing request for getting versions: %w", err)
+		return
+	}
+
+	resp, err := aac.c.Do(req)
+	if err != nil {
+		err = fmt.Errorf("executing http request %v: %w", req, err)
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("%d: %w", resp.StatusCode, ErrBadHTTPStatusCode)
+		return
+	}
+
+	respBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		err = fmt.Errorf("reading response body: %w", err)
+		return
+	}
+	builds = new(VersionBuilds)
+	err = json.Unmarshal(respBytes, builds)
+
+	if err != nil {
+		err = fmt.Errorf("unmarshaling build list: %w", err)
+		return
+	}
+
+	return builds, nil
+}
+
+func (aac ArtifactAPIClient) GetBuildDetails(ctx context.Context, version string, buildID string) (buildDetails *BuildDetails, err error) {
+	joinedURL, err := url.JoinPath(aac.apiURL, fmt.Sprintf(artifactAPIV1BuildDetailsEndpoint, version, buildID))
+	if err != nil {
+		err = fmt.Errorf("composing URL with %q %q: %w", aac.apiURL, fmt.Sprintf(artifactAPIV1BuildDetailsEndpoint, version, buildID), err)
+		return
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, joinedURL, nil)
+
+	if err != nil {
+		err = fmt.Errorf("composing request for getting build details: %w", err)
+		return
+	}
+
+	resp, err := aac.c.Do(req)
+	if err != nil {
+		err = fmt.Errorf("executing http request %v: %w", req, err)
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("%d: %w", resp.StatusCode, ErrBadHTTPStatusCode)
+		return
+	}
+
+	respBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		err = fmt.Errorf("reading response body: %w", err)
+		return
+	}
+	buildDetails = new(BuildDetails)
+	err = json.Unmarshal(respBytes, buildDetails)
+
+	if err != nil {
+		err = fmt.Errorf("unmarshaling build details: %w", err)
+		return
+	}
+
+	return buildDetails, nil
+}
+
+type ArtifactAPIClientOpt func(aac *ArtifactAPIClient)
+
+func WithUrl(url string) ArtifactAPIClientOpt {
+	return func(aac *ArtifactAPIClient) { aac.apiURL = url }
+}
+
+func WithHttpClient(client httpDoer) ArtifactAPIClientOpt {
+	return func(aac *ArtifactAPIClient) { aac.c = client }
+}
+
+func NewArtifactAPIClient(opts ...ArtifactAPIClientOpt) *ArtifactAPIClient {
```

Comment:
It's pretty much personal preference, but I find it easier to quickly read the code when the `NewWHATEVER` is the first function in the file.